scalyr_agent/log_processing.py

```diff
@@ -88,45 +88,150 @@
-        return None, None, None, None
+        # parse the stream type
+        stream = parts[1]
+        if stream != "stdout" and stream != "stderr":
+            raise CRIParseError("Unknown stream type", line)
 
-    # skip over the ' ' delimiter and parse the tags
-    line = line[index + 1 :]
-    index = line.find(" ")
-    if index < 0:
-        return None, None, None, None
+        tags = parts[2].split(":")
+        line = parts[3] if len(parts) == 4 else ""
 
-    tags = line[:index]
+        return timestamp, stream, tags, line, raw_timestamp
 
-    # skip over the ' ' delimiter and get the log line
-    line = line[index + 1 :]
+class CRILogLineBuffer(object):
+    def __init__(self, first_line_time):
+        self.__stream = None
+        self.__lines = []
+        self.__timestamp = None
+        self.__raw_timestamp = None
+        self.__is_partial = True
+        self.__first_line_time = first_line_time
 
-    return timestamp, stream, tags, line
+    def append(self, log_line):
+        if not self.__lines:
+            self.__timestamp = log_line.timestamp
+            self.__raw_timestamp = log_line.raw_timestamp
+            self.__stream = log_line.stream
 
+        self.__lines.append(log_line.message)
+
+        if log_line.is_full:
+            self.__is_partial = False
+
+    @property
+    def stream(self):
+        return self.__stream
+
+    @property
+    def first_line_time(self):
+        return self.__first_line_time
+
+    @property
+    def is_partial(self):
+        return self.__is_partial
+
+    def payload_length(self):
+        return sum(map(len, self.__lines))
+
+    def build_cri_result(self, include_raw_timestamp_field):
+        def strip_new_line(s):
+            if s and s[-1] == "\n":
+                return s[:-1]
+            return s
+
+        lines_string = "".join(map(strip_new_line, self.__lines)) + "\n"
+
+        result = LogLine(lines_string.encode("utf-8"))
+        result.timestamp = self.__timestamp
+        result.attrs = {"stream": self.__stream}
+        if include_raw_timestamp_field:
+            result.attrs["raw_timestamp"] = self.__raw_timestamp
+
+        return result
+
+class CRIStreamBuffers(object):
+    def __init__(self, max_line_size, line_completion_wait_time):
+        self.__max_line_size = max_line_size
+        self.__line_completion_wait_time = line_completion_wait_time
+        self.__buffers = {}
+
+    def append_log_line(self, log_line, current_time):
+        stream = log_line.stream
+        if stream not in self.__buffers:
+            self.__buffers[stream] = CRILogLineBuffer(current_time)
+
+        self.__buffers[stream].append(log_line)
+
+    def __buffer_completed(self, buffer, current_time):
+        return (
+                not buffer.is_partial
+                or current_time - buffer.first_line_time > self.__line_completion_wait_time
+                or buffer.payload_length() >= self.__max_line_size
```

Comment:
Do we today, some how indicate is a line has been stripped due to `max_line_size` config option? E.g. by appending `...` to the end of the line or similar.

I know we do that on the server side, but not sure of about the client / agent side.

Reply:
The line is not stripped but split, the rest is sent in the next message.
Do you think it would make sense to indicate that somehow?
i.e.:
message_part 1 ....
... message_part 2

How is this handled on the server side?

Reply:
On the server, we just truncate the field value and append `...` to the values which are over the server side account level defined limit (so we don't split it across multiple lines).

In case we split it and not truncate it, I think it's fine to keep that logic as-is.